src/matrix_construction.cpp

```diff
@@ -130,18 +144,65 @@ List rcpp_h_eval(int k, NumericVector xd, NumericVector x, IntegerVector col_idx
   }
 
   // Now construct H in sparse triplet format
-  IntegerVector i_vec (N);
-  IntegerVector j_vec (N);
-  NumericVector x_vec (N);
+  std::vector<T> h_list;
+  h_list.reserve(N);
   int l = 0;
   for (int j = 0; j < n_col; j++) {
     for (int i = I[j]; i < n_row; i++) {
-      i_vec[l] = i;
-      j_vec[l] = j;
-      x_vec[l] = hxj(k, xd, x[i], col_idx[j]);
+      h_list.push_back(T(i, j, hxj(k, xd, x[i], col_idx[j])));
       l++;
     }
   }
 
-  return List::create(Named("i") = i_vec, Named("j") = j_vec, Named("x") = x_vec);
+  SparseMatrix<double> h_mat(n_row, n_col);
+  h_mat.setFromTriplets(h_list.begin(), h_list.end());
+  return h_mat;
+}
+
+/******************************************************************************/
+// Evaluation of discrete spline basis at arbitrary query points
+
+// [[Rcpp::export]]
+Eigen::SparseMatrix<double> rcpp_n_eval(int k, NumericVector xd, NumericVector x, bool normalized, IntegerVector knot_idx) {
+  int n_xd = xd.size();
+  int n_knot_idx = knot_idx.size();
+  double max_diff = Rcpp::max(xd[Rcpp::seq(1, n_xd-1)]-xd[Rcpp::seq(0, n_xd-2)]);
+  IntegerVector ext_knot_idx(n_knot_idx+k+1);
+  NumericVector ext_xd(n_xd+k+1);
+  ext_knot_idx[Rcpp::seq(0, n_knot_idx-1)] = knot_idx;
```

Comment:
Should we move this to always occur inside `rcpp_n_mat` and remove the corresponding code from the R `n_mat` function?  Would anybody ever want to call `rcpp_n_mat` _without_ the extra knots appended?